gadget/validate.go

```diff
@@ -35,6 +35,45 @@ type validationState struct {
 	SystemSave *VolumeStructure
 }
 
+type ValidationConstraints struct {
+	// EncryptedData when true indicates that the gadget will be used on a
+	// device where the data partition will be encrypted.
+	EncryptedData bool
+}
+
+// Validate validates the given gadget metadata against the consistency rules
+// for roles, labels etc as implied by the model and extra constraints
+// tha might be known only at runtime.
+func Validate(info *Info, model Model, extra *ValidationConstraints) error {
+	if err := ruleValidateVolumes(info.Volumes, model); err != nil {
+		return err
+	}
+	if extra != nil {
+		if extra.EncryptedData {
+			if err := validateEncryptionSupport(info); err != nil {
+				return fmt.Errorf("gadget does not support encrypted data: %v", err)
+			}
+		}
+	}
+	return nil
+}
+
+func validateEncryptionSupport(info *Info) error {
+	for name, vol := range info.Volumes {
+		var haveSave bool
+		for _, s := range vol.Structure {
+			if s.Role == SystemSave {
+				haveSave = true
+			}
+		}
+		if !haveSave {
+			return fmt.Errorf("volume %q has no structure with system-save role", name)
+		}
+		// XXX: shall we make sure that size of ubuntu-save is reasonable?
```

Comment:
thinking about it, maybe this should be a TODO:UC20: ? or a general TODO ?

Reply:
this wasn't new here but I'm turning it into a TODO:UC20, though we probably can't address it for a little while